modules/core/04-channel/keeper/grpc_query.go

```diff
@@ -335,8 +335,18 @@ func (q Keeper) PacketAcknowledgements(c context.Context, req *types.QueryPacket
 			return err
 		}
 
-		ack := types.NewPacketState(req.PortId, req.ChannelId, sequence, value)
-		acks = append(acks, &ack)
+		if len(req.Commitments) > 0 {
+			for _, commitment := range req.Commitments {
+				if sequence == commitment.Sequence {
+					ack := types.NewPacketState(req.PortId, req.ChannelId, sequence, value)
+					acks = append(acks, &ack)
+				}
+			}
```

Comment:
You'll want to actually modify the query being done. Since we can assume there to be thousands of acknowledgements written in state, a small pagination limit might actually return the incorrect values since the pagination might stop before getting all the necessary commitments

What we want to do is query for an acknowledgement for each commitment provided. This will look similar to this [code](https://github.com/cosmos/ibc-go/pull/375/files#diff-74764e734e2ec27e29dc1f3e36deef94d0c1fa798c60478cda09f2cd9a1316b4L384) but for acknowledgements

Reply:
Ah okay! So querying for specific packet acks should be done outside the bounds of the paginated query.
I think the link you provided is pointing to tests in this PR, after taking a look I assume it's this code you were referencing (correct me if not):

```
for i, seq := range req.PacketCommitmentSequences {
		if seq == 0 {
			return nil, status.Errorf(codes.InvalidArgument, "packet sequence %d cannot be 0", i)
		}

		// if packet receipt exists on the receiving chain, then packet has already been received
		if _, found := q.GetPacketReceipt(ctx, req.PortId, req.ChannelId, seq); !found {
			unreceivedSequences = append(unreceivedSequences, seq)
		}
}
```

Reply:
yup :+1: . Since the caller provides a list of sequences we don't need the pagination. We still need the pagination if the caller passes in an empty list of sequences (keep old behaviour)

Reply:
Updated! 

I generally prefer returning early than having most of the function body wrapped in an if/else. Also omitted the following, but let me know if it's required.

```
if seq == 0 {
	return nil, status.Errorf(codes.InvalidArgument, "packet sequence %d cannot be 0", i)
}
```

Reply:
I think it is fine to omit this, it shouldn't cause any issues